src/opaque_closure.c

```diff
@@ -0,0 +1,69 @@
+#include "julia.h"
+#include "julia_internal.h"
+
+JL_DLLEXPORT jl_value_t *jl_invoke_opaque_closure(jl_opaque_closure_t *oc, jl_value_t **args, size_t nargs)
+{
+    jl_value_t *argt = jl_tparam0(jl_typeof(oc));
+    if (!jl_tupletype_length_compat(argt, nargs))
+        jl_error("Incorrect argument count for OpaqueClosure");
+    argt = jl_unwrap_unionall(argt);
+    assert(jl_is_datatype(argt));
+    jl_svec_t *types = jl_get_fieldtypes((jl_datatype_t*)argt);
+    size_t ntypes = jl_svec_len(types);
+    for (int i = 0; i < nargs; ++i) {
+        jl_typeassert(args[i], i > ntypes ?
+            jl_unwrap_vararg(jl_svecref(types, ntypes-1)) : jl_svecref(types, i));
+    }
+    jl_value_t *ret = NULL;
+    JL_GC_PUSH1(&ret);
+    jl_ptls_t ptls = jl_get_ptls_states();
+    size_t last_age = ptls->world_age;
```

Comment:
I think you want to move everything above here into `jl_f_opaque_closure_call`, so that `invoke` isn't expected to check these conditions. (later, I think we can teach the method cache lookup to do this as part of the dispatch computation and bypass `jl_f_opaque_closure_call` when creating new cache entries).